Show number of days and players on Commons Overview

mfletcher2 opened the pull request:
This PR adds the display of the number of days since the start and the number of players. The day count is computed with the start date and the current date on the client with the new function `daysSinceTimestamp` in `dateUtils`. The number of players is retrieved from a new endpoint to get a single `CommonsPlus`. It contains information about the number of players and cows in addition to the other `Commons` fields. For testing purposes, the player delete endpoint now returns a message that says the new number of players.

# Screenshots
Before:
![Screenshot_20230531_231445](https://github.com/ucsb-cs156-s23/proj-happycows-s23-6pm-1/assets/92177225/31ee9f3c-e734-4751-8c36-6700480e84de)
After:
![Screenshot_20230531_231512](https://github.com/ucsb-cs156-s23/proj-happycows-s23-6pm-1/assets/92177225/dd9e541a-bfb8-4c91-9cac-588a572b6484)


# Storybook
[Before](https://ucsb-cs156-s23.github.io/proj-happycows-s23-6pm-1/storybook/?path=/story/components-commons-commonsoverview--uncontrolled)
[After](https://ucsb-cs156-s23.github.io/proj-happycows-s23-6pm-1/prs/38/storybook/?path=/story/components-commons-commonsoverview--uncontrolled)
(The counters appear to be working on the before storybook because they were given values in the fixtures, but not implemented in the rest of the code)

Closes #28

mfletcher2 committed:
mf - Add function to calculate days since start

mfletcher2 committed:
mf - Show day and player count on commons overview

mfletcher2 committed:
mf - Fix error when User List is null

mfletcher2 committed:
mf - Use existing getNumUsers function
to avoid null issues

mfletcher2 committed:
mf - Fix CommonsController mutation tests
by removing an unnecessary update for numPlayers and testing the correct value of numPlayers

mfletcher2 committed:
mf - Update player count if user is deleted
This is only possible through the API AFAIK

mfletcher2 committed:
mf - Fix mutation test for user delete operation
by changing the response message to include the current number of users

mfletcher2 committed:
Merge remote-tracking branch 'origin/main' into max-commonsCounter

mfletcher2 committed:
mf - Change totalPlayers to numPlayers for story

mfletcher2 committed:
mf - Fix spacing in CommonsControllerTests

mfletcher2 committed:
mf - Fix mutation tests for the day counter

GlassBricks commented:
The handling of the day number is good.

However, I think the number of players should be computed (by the backend) instead of a manually edited field in `Commons`. It definetely doesn't make sense to have it in `CreateCommonsParams`.
Instead, there's already a backend thing that includes this computed info (`CommonsPlus`, and the endpoint `commons/allplus`); we could add another endpoint to get info for a specific commons, and use that in the frontend.

GlassBricks reviewed (changes requested):
The handling of the day number is good.

However, I think the number of players should be computed (by the backend) instead of a manually edited field in `Commons`. It definitely doesn't make sense to have it in `CreateCommonsParams`.
There's already a backend thing that includes this computed info (`CommonsPlus`, and the endpoint `commons/allplus`); we could add another endpoint to get info for a specific commons, and use that in the frontend.

mfletcher2 committed:
mf - wip: remove numPlayers variable

mfletcher2 committed:
mf - Add endpoint to get CommonsPlus by id
and use this endpoint in the commons overview page instead of /get/commons
and also fix tests to work with this new parameter

mfletcher2 committed:
mf - Add backend tests for GET commonsPlus

mfletcher2 committed:
mf - Fix storybook page not using CommonsPlus

GlassBricks commented on frontend/src/main/pages/PlayPage.js:
```diff
@@ -47,6 +47,20 @@ export default function PlayPage() {
     );
```
We don't need another backend GET for normal commons; you can use `commonsPlus.commons` instead, avoiding an extra query

mfletcher2 committed:
mf - Remove unnecessary call to GET Commons

GlassBricks commented on frontend/src/main/pages/PlayPage.js:
```diff
@@ -143,7 +128,7 @@ export default function PlayPage() {
           {!!userCommons &&
             <CardGroup >
-              <ManageCows userCommons={userCommons} commons={commons} onBuy={onBuy} onSell={onSell} />
+              <ManageCows userCommons={userCommons} onBuy={onBuy} onSell={onSell} />
```
I think you still need `commons={commonsPlus.commons}` here... now market cow price isn't being shown

mfletcher2 replied:
Thanks for catching that

mfletcher2 committed:
mf - fix cow price not shown

GlassBricks reviewed (dismissed):
LGTM

pconrad commented on frontend/src/main/pages/PlayPage.js:
```diff
@@ -32,20 +32,19 @@ export default function PlayPage() {
       }
     );
-  // Stryker enable all 
+  // Stryker enable all
```
Should be 
```suggestion
  // Stryker restore all
```

Please see: https://ucsb-cs156-s23.slack.com/archives/C050779M19P/p1685726928710669

pconrad commented on frontend/src/main/utils/dateUtils.js:
```diff
@@ -6,4 +6,15 @@ const timestampToDate = (timestamp) => {
 
-export {timestampToDate, padWithZero};
\ No newline at end of file
+// Disabling stryker here because the returned value depends on the date the test is run
+// so a deterministic test is not possible
+// Stryker disable all
```
There's a better approach to this: mock the system time:

```js
jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));
```

See also:  https://stackoverflow.com/questions/29719631/how-do-i-set-a-mock-date-in-jest/63377110#63377110

pconrad commented on frontend/src/main/utils/dateUtils.js:
```diff
@@ -6,4 +6,15 @@ const timestampToDate = (timestamp) => {
+    var timeDiff = Math.abs(today.getTime() - startingDate.getTime());
+    return Math.ceil(timeDiff / (1000 * 3600 * 24));
+}
+// Stryker enable all
```
we can probably eliminate the need for a stryker disable here, so we can remove this; but if we don't, it should be

```suggestion
// Stryker restore all
```

pconrad reviewed (changes requested):
Good overall, but there are a few things that need to be fixed.

mfletcher2 committed:
mf - Stryker enable to restore on PlayPage

mfletcher2 committed:
mf - Add test for daysSinceTimestamp

mfletcher2 committed:
mf - Fix newly failing mutation tests

mfletcher2 committed:
mf - Fix remaining mutation test

pconrad reviewed (approved):
LGTM
